test/e2e/network_policy.go

```diff
@@ -0,0 +1,630 @@
+		testCanConnect(f, ns, "client-b", service, 81)
+	})
+
+	It("should enforce policy based on NamespaceSelector [Feature:NetworkPolicy]", func() {
+		nsA := f.Namespace
+		nsBName := f.BaseName + "-b"
+		// The CreateNamespace helper uses the input name as a Name Generator, so the namespace itself
+		// will have a different name than what we are setting as the value of ns-name.
+		// This is fine as long as we don't try to match the label as nsB.Name in our policy.
+		nsB, err := f.CreateNamespace(nsBName, map[string]string{
+			"ns-name": nsBName,
+		})
+		Expect(err).NotTo(HaveOccurred())
+		setNamespaceIsolation(f, nsA, "DefaultDeny")
+
+		// Create Server with Service in NS-B
+		By("Creating a webserver tied to a service.")
+		serverPod, service := createServerPodAndService(f, nsA, "server", []int{80})
+		defer func() {
+			By("Cleaning up the server.")
+			if err := f.ClientSet.Core().Pods(nsA.Name).Delete(serverPod.Name, nil); err != nil {
+				framework.Failf("unable to cleanup pod %v: %v", serverPod.Name, err)
+			}
+		}()
+		defer func() {
+			By("Cleaning up the server's service.")
+			if err := f.ClientSet.Core().Services(nsA.Name).Delete(service.Name, nil); err != nil {
+				framework.Failf("unable to cleanup svc %v: %v", service.Name, err)
+			}
+		}()
+		framework.Logf("Waiting for server to come up.")
+		err = framework.WaitForPodRunningInNamespace(f.ClientSet, serverPod)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Create Policy for that service that allows traffic only via namespace B
+		By("Creating a network policy for the server which allows traffic from namespace-b.")
+		policy, err := f.ClientSet.Extensions().NetworkPolicies(nsA.Name).Create(&extensions.NetworkPolicy{
+			ObjectMeta: api.ObjectMeta{
+				Name: "allow-ns-b-via-namespace-selector",
+			},
+			Spec: extensions.NetworkPolicySpec{
+				// Apply to server
+				PodSelector: unversioned.LabelSelector{
+					MatchLabels: map[string]string{
+						"pod-name": serverPod.Name,
+					},
+				},
+				// Allow traffic only from NS-B
+				Ingress: []extensions.NetworkPolicyIngressRule{{
+					From: []extensions.NetworkPolicyPeer{{
+						NamespaceSelector: &unversioned.LabelSelector{
+							MatchLabels: map[string]string{
+								"ns-name": nsBName,
+							},
+						},
+					}},
+				}},
+			},
+		})
+		Expect(err).NotTo(HaveOccurred())
+		defer func() {
+			By("Cleaning up the policy")
+			if err := f.ClientSet.Extensions().NetworkPolicies(nsA.Name).Delete(policy.Name, nil); err != nil {
+				framework.Failf("unable to cleanup policy %v: %v", policy.Name, err)
+			}
+		}()
+
+		testCannotConnect(f, nsA, "client-a", service, 80)
+		testCanConnect(f, nsB, "client-b", service, 80)
+	})
+})
+
+func testCanConnect(f *framework.Framework, ns *api.Namespace, podName string, service *api.Service, targetPort int) {
+	podClient := createClientPod(f, ns, podName, service, targetPort)
+	By("Creating client that should successfully connect to the opened port.")
```

Comment:
This "By" isn't great - it might not be a specific port that is opened.

Would be more clear if it was:

```
"Creating client pod %s that should successfully connect to %s", podName, service.Name)
```